primitives/bls-primitives/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+		CryptoTypePublicPair(CRYPTO_ID, self.0.to_vec())
+	}
+}
+
+impl AsRef<[u8]> for Signature {
+	fn as_ref(&self) -> &[u8] {
+		self.0.as_ref()
+	}
+}
+
+#[cfg(feature = "std")]
+impl sp_core::crypto::Pair for Pair {
+	type Public = Public;
+	type Seed = Seed;
+	type Signature = Signature;
+	type DeriveError = Error;
+
+	fn generate_with_phrase(password: Option<&str>) -> (Self, String, Self::Seed) {
+		let mnemonic = Mnemonic::new(MnemonicType::Words24, Language::English);
+		let phrase = mnemonic.phrase();
+		let (pair, seed) = Self::from_phrase(phrase, password)
+			.expect("All phrases generated by Mnemonic are valid; qed");
+		(pair, phrase.to_owned(), seed)
+	}
+
+	fn from_phrase(
+		phrase: &str,
+		password: Option<&str>,
+	) -> Result<(Pair, Seed), SecretStringError> {
+		let big_seed = seed_from_entropy(
+			Mnemonic::from_phrase(phrase, Language::English)
+				.map_err(|_| SecretStringError::InvalidPhrase)?
+				.entropy(),
+			password.unwrap_or(""),
+		)
+		.map_err(|_| SecretStringError::InvalidSeed)?;
+		let mut seed = Seed::default();
+		seed.copy_from_slice(&big_seed[0..32]);
+		let secret = SecretKey::key_gen(&seed, &[]).unwrap();
+		let pair = Pair { public: secret.sk_to_pk().compress().into(), secret };
+		Ok((pair, seed))
+	}
+
+	#[cfg(feature = "std")]
+	fn derive<Iter: Iterator<Item = DeriveJunction>>(
+		&self,
+		path: Iter,
+		seed: Option<Self::Seed>,
+	) -> Result<(Self, Option<Self::Seed>), Self::DeriveError> {
+		if seed.is_none() {
+			return Err(Error::InvalidSeed)
+		}
+		let mut master_key = SecretKey::key_gen(&seed.unwrap(), &[])?;
+		for junction in path {
+			let index_bytes = [
+				junction.inner()[0],
+				junction.inner()[1],
+				junction.inner()[2],
+				junction.inner()[3],
+			];
+			master_key = master_key.derive_child_eip2333(u32::from_be_bytes(index_bytes))
+		}
+		Ok((Pair { public: master_key.sk_to_pk().compress().into(), secret: master_key }, seed))
+	}
+
+	fn from_seed(seed: &Self::Seed) -> Self {
+		Self::from_seed_slice(&seed[..]).expect("seed needs to be of valid length; qed")
+	}
+
+	fn from_seed_slice(seed: &[u8]) -> Result<Self, SecretStringError> {
+		println!("Seed: {:?}, len: {:?}", seed, seed.len());
+		let secret = match SecretKey::key_gen(seed, &[]) {
+			Ok(secret) => secret,
+			Err(err) => {
+				println!("BLS err: {:?}", err);
+				return Err(SecretStringError::InvalidSeed)
+			},
+		};
+
+		Ok(Pair { public: secret.sk_to_pk().compress().into(), secret })
+	}
+
+	fn sign(&self, message: &[u8]) -> Self::Signature {
+		self.secret.sign(message, DST.as_ref(), &[]).into()
+	}
+
+	fn verify<M: AsRef<[u8]>>(sig: &Self::Signature, message: M, pubkey: &Self::Public) -> bool {
+		let pubkey = PublicKey::from_bytes(&pubkey.0).expect("Expected valid public key");
+		let signature =
+			BLSSignature::from_bytes(sig.0.as_ref()).expect("Expected valid BLS signature");
+
+		signature.verify(
+			true,
+			message.as_ref(),
+			DST.as_ref(),
+			&[], // TODO: wtf is this?
```

Comment:
Still pending

Reply:
We need to handle it outside of this PR.